gpu-direct/ubuntu/entrypoint.sh

```diff
@@ -2,23 +2,51 @@
 MOFED=/run/mellanox/drivers
 NVIDIA=/run/nvidia/drivers
 KERNEL_VERSION=$(uname -r)
-# Install linux headers
-apt-get -yq install linux-headers-${KERNEL_VERSION}
-# Patch filesystem with components from both Mellanox and Nvidia Drivers
-ln -sf ${MOFED}/usr/src/ofa_kernel /usr/src/ofa_kernel
-ln -sf ${NVIDIA}/usr/src/nvidia-* /usr/src/.
-touch /lib/modules/${KERNEL_VERSION}/modules.order
-touch /lib/modules/${KERNEL_VERSION}/modules.builtin
-mkdir -p /lib/modules/${KERNEL_VERSION}/updates/dkms
-ln -sf ${MOFED}/lib/modules/${KERNEL_VERSION}/updates/dkms/* /lib/modules/${KERNEL_VERSION}/updates/dkms/
-ln -sf ${NVIDIA}/lib/modules/${KERNEL_VERSION}/updates/dkms/* /lib/modules/${KERNEL_VERSION}/updates/dkms/
-ln -sf ${MOFED}/var/lib/dkms/mlnx-ofed-kernel /var/lib/dkms/mlnx-ofed-kernel
-ln -sf ${NVIDIA}/var/lib/dkms/nvidia /var/lib/dkms/nvidia
-mkdir -p /etc/infiniband
-cp /root/nv_peer_memory/nv_peer_mem.conf /etc/infiniband/
-# Build NV PEER MEMORY module
-cd /root/nv_peer_memory
-make clean && make && make install
-./nv_peer_mem restart
-./nv_peer_mem status
-sleep infinity
+
+function exit_on_error() {
+    $@
+    if [[ $? -ne 0 ]]; then
+        echo "Error occured while executing: $1"
+        exit 1
+    fi
+}
+
+function install_prereq_runtime() {
+    # Install linux headers
+    apt-get -yq install linux-headers-${KERNEL_VERSION}
+}
+
+function prepare_build_env() {
+    # Patch filesystem with components from both Mellanox and Nvidia Drivers
+    ln -sf ${MOFED}/usr/src/ofa_kernel /usr/src/ofa_kernel
+    ln -sf ${NVIDIA}/usr/src/nvidia-* /usr/src/.
+    touch /lib/modules/${KERNEL_VERSION}/modules.order
+    touch /lib/modules/${KERNEL_VERSION}/modules.builtin
+    mkdir -p /lib/modules/${KERNEL_VERSION}/updates/dkms
+    ln -sf ${MOFED}/lib/modules/${KERNEL_VERSION}/updates/dkms/* /lib/modules/${KERNEL_VERSION}/updates/dkms/
+    ln -sf ${NVIDIA}/lib/modules/${KERNEL_VERSION}/updates/dkms/* /lib/modules/${KERNEL_VERSION}/updates/dkms/
+    ln -sf ${MOFED}/var/lib/dkms/mlnx-ofed-kernel /var/lib/dkms/mlnx-ofed-kernel
+    ln -sf ${NVIDIA}/var/lib/dkms/nvidia /var/lib/dkms/nvidia
+    mkdir -p /etc/infiniband
+    cp /root/nv_peer_memory/nv_peer_mem.conf /etc/infiniband/
+}
+
+function build_modules() {
+    # Build NV PEER MEMORY module
+    cd /root/nv_peer_memory
+    make clean && make && make install
+    if [[ $? -ne 0 ]]; then
+        return 1
+    fi
+    ./nv_peer_mem stop
+    ./nv_peer_mem start
+    if [[ $? -ne 0 ]]; then
+        return 1
+    fi
+    ./nv_peer_mem status
+}
+
+exit_on_error install_prereq_runtime
+exit_on_error prepare_build_env
```

Comment:
you will check just the last error

Reply:
what do you mean ? 
exit_on_error will exit the shell

(not a bash expert so i might have gotten things wrong here)

Reply:
I mean the prepare_build_env run several commands so the return value of the function will be the last command. If a command will failed in the middle like  touch /lib/modules/${KERNEL_VERSION}/modules.builtin and the last command of will pass it won't failed.

Reply:
ah yes, but then build_modules would fail to compile.
from the logs we would understand what failed.

my main concern tbh is the module build phase and the module load phase.

for completness shall i add  a return_on_error method ? and use it for each call ?

Reply:
yes

Reply:
that not what I meant. I wanted you to check the error code per each command that you do.